Added initial version of decoder

thelimar opened the pull request:
Not tested lol

thelimar committed:
Added initial version of decoder

VasiliyMatr commented on instr/gen_instr_id.py:
```diff
@@ -1,22 +1,35 @@
+    gen = open("instr_id.gen.hpp", "w+")
+
+    gen.write(write_buffer)
+
```
Add clang-format run on generated file, please

thelimar replied:
Added

VasiliyMatr commented:
There are some generation errors: misspelled 'switch'; forgotten bit:: namespace when using bit operations; 'Word' type is not defined in our project, etc. Please, fix those and make sure that project builds locally

UPD: closed

VasiliyMatr commented:
Debug masks are currently used to decode instructions. It seems wrong because those are added for debugging. Thus there can be hard-to-find errors in decoding.
There are two options:
1) Consider using 'decodertree' from risc-v.yaml to perform decoding
2) Use other instructions describtion if you want masks decoding

UPD: closed

thelimar committed:
added decoder tree decoding and tests

TheRedHotHabanero commented on instr/tests/src/test_instr_decoding.cpp:
```diff
@@ -0,0 +1,85 @@
+
+TEST(Instr, add)
+{
+    Instr test = Instr(static_cast<InstrCode>(0x00710a13)); //ADDI x20, x2, 7
```
its better to use common comment style, minor

thelimar replied:
fixed

TheRedHotHabanero commented on instr/tests/src/test_instr_decoding.cpp:
```diff
@@ -0,0 +1,85 @@
+
+TEST(Instr, garbage)
+{
+    Instr test = Instr(static_cast<InstrCode>(0xfffa056f)); //hz, absolutno
```
hehe or not hehe? please improve readabilty

thelimar replied:
removed in last commit

TheRedHotHabanero commented on instr/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+    sim::instr
+)
+
+target_sources(test_instr PRIVATE src/main.cpp src/test_instr_decoding.cpp ../generated/src/decode.gen.cpp)
```
please take a look at error with gmock include in the next target

thelimar replied:
cmake seems to work fine

VasiliyMatr commented on instr/gen_decode.py:
```diff
@@ -1,14 +1,137 @@
+    return write_buffer
+
+
+def GenerateSwitchCaseMasks(yaml_dump: dict) -> str:
```
Remove unused code, please

thelimar replied:
This code might be used later for perf tests

VasiliyMatr commented:
Please, get rid off:
1) Trailing whitespaces
2) Old code in comments

UPD: closed

VasiliyMatr commented on instr/tests/src/test_instr_decoding.cpp:
```diff
@@ -0,0 +1,85 @@
+
+TEST(Instr, garbage)
+{
+    Instr test = Instr(static_cast<InstrCode>(0xfffa056f)); //hz, absolutno
```
Please, improve this test:
Iterate over all uint32_t values. Program must end successfully

thelimar replied:
added such test, takes long time though

VasiliyMatr commented on instr/gen_decode.py:
```diff
@@ -1,14 +1,137 @@
+        "namespace sim {" "\n"                      +\
+        "namespace instr {" "\n"                    +\
+        "\n"                                        +\
+        "Instr::Instr(InstrCode instr_code) {" "\n"
```
Maybe, it will be more efficient to always assign rd, rs1 and rs2 here (and remove assigning from decode tree). This will reduce code size.

VasiliyMatr replied:
Such changes require perf tests. Thus, i advice to deal with this later

thelimar replied:
ok, might address later

thelimar committed:
fixed bugs and tests

VasiliyMatr committed:
[instr] MR fixes

- Fix formatting
- Remove old code
- Fix instr/CMakeLists.txt codestyle
- Fix decoder garbage test
- Add a few comments

VasiliyMatr committed:
[common][hotfix] Remove deprecated common code

- Remove deprecated common::bit:: functions
- Fix bit::signExtend UB on highest bit case
- Port instr:: module to new common::bit:: functions
- A few other fixes

VasiliyMatr reviewed (approved):
Can merge
